cql3: prepare selectors

avikivity opened the pull request:
CQL statements carry expressions in many contexts: the SELECT, WHERE, SET, and IF clauses, plus various attributes. Previously, each of these contexts had its own representation for an expression, and another one for the same expression but before preparation. We have been gradually moving towards a uniform representation of expressions.

This series tackles SELECT clause elements (selectors), in their unprepared phase. It's relatively simple since there are only five types of expression components (column references, writetime/ttl modifiers, function calls, casts, and field selections). Nevertheless, there isn't much commonality with previously converted expression elements so quite a lot of code is involved.

After the series, we are still left with a custom post-prepare representation of expressions. It's quite complicated since it deals with two passes, for aggregation, so it will be left for another series.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1783/

avikivity commented:
v2: workaround clang 15 refusing to capture bindings in lambdas.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1784/

nyh commented on cql3/expr/prepare_expr.cc:
```diff
@@ -810,11 +810,29 @@ cast_prepare_expression(const cast& c, data_dictionary::database db, const sstri
 
+static
+std::optional<expression>
+sql_cast_prepare_expression(const cast& c, data_dictionary::database db, const sstring& keyspace, const schema* schema_opt, lw_shared_ptr<column_specification> receiver) {
```
We should eventually document the difference between what you called here "C" and "SQL" type casts (in a document and/or a comment). Or do we have such documentation already? Are these names you invented now or we used them somewhere before?

In `test/cql-pytest/test_cast*.py` I called these casts "query" and "data" casts respectively - what I called "query" casts can only change, losslessly, constants appearing in the query, whereas the "data" casts can do lossy conversions of data read from the database.

avikivity replied:
The user-facing documentation calls the SQL-style cast "casting" and doesn't mention the C-style cast.

I don't propose we use these terms in user-facing documentation.

nyh commented on cql3/expr/prepare_expr.cc:
```diff
@@ -997,7 +997,7 @@ prepare_function_call(const expr::function_call& fc, data_dictionary::database d
         .lwt_cache_id = fc.lwt_cache_id
     };
-    if (all_terminal && fun->is_pure() && !fun->is_aggregate()) {
+    if (all_terminal && fun->is_pure() && !fun->is_aggregate() && !fun->requires_thread()) {
```
I guess it's ok to skip an optimization, but just a thought: It's a bit odd that if the function is particularly heavy (I am guessing functions that want to run in a thread are by average heavier), then we decide not to optimize the number of calls to it.

avikivity replied:
It would be nice to perform the optimization, but note we don't right now, so this isn't a regression.

nyh commented on cql3/expr/prepare_expr.cc:
```diff
@@ -1034,7 +1034,7 @@ test_assignment_function_call(const cql3::expr::function_call& fc, data_dictiona
 static assignment_testable::test_result expression_test_assignment(const data_type& expr_type,
                                                                    const column_specification& receiver) {
-    if (receiver.type->underlying_type() == expr_type->underlying_type()) {
+    if (receiver.type->underlying_type() == expr_type->underlying_type() || (receiver.type == long_type && expr_type->is_counter())) {
```
I'm not sure that I understood you correctly, but are you saying that before this patch sum(counter) didn't work and after the patch, it does? Can we have a test for this case (and perhaps even an issue about it)?

avikivity replied:
At this point in the series, this code isn't called with aggregate functions.

nyh commented on cql3/selection/selectable.cc:
```diff
@@ -125,32 +104,13 @@ selectable::with_field_selection::new_selector_factory(data_dictionary::database
-sstring
-selectable::with_cast::to_string() const {
-    return format("cast({} as {})", _arg->to_string(), cql3_type(_type).to_string());
+    return format("{}.{}", _selected->to_string(), "blarf");
```
blarf??

nyh commented on cql3/selection/selectable.cc:
```diff
@@ -76,27 +76,6 @@ selectable::writetime_or_ttl::to_string() const {
     return format("{}({})", _is_writetime ? "writetime" : "ttl", _id->to_string());
 }
 
-shared_ptr<selector::factory>
```
I don't know where to ask this so I'll ask here:
If we prepare a statement which mentions a UDF name, and later we drop this UDF and then try to execute the prepared statement - what will happen? Do we have a test for this case?

avikivity replied:
We should drop the function from the cache. We probably don't, the cached statement will hold on to the shared pointer and keep it alive until it's evicted.

avikivity replied:
```c++
void query_processor::migration_subscriber::on_drop_keyspace(const sstring& ks_name) {
    remove_invalid_prepared_statements(ks_name, std::nullopt);
}

void query_processor::migration_subscriber::on_drop_column_family(const sstring& ks_name, const sstring& cf_name) {
    remove_invalid_prepared_statements(ks_name, cf_name);
}

void query_processor::migration_subscriber::on_drop_user_type(const sstring& ks_name, const sstring& type_name) {
}

void query_processor::migration_subscriber::on_drop_function(const sstring& ks_name, const sstring& function_name) {
    log.warn("{} event ignored", __func__);
}

void query_processor::migration_subscriber::on_drop_aggregate(const sstring& ks_name, const sstring& aggregate_name) {
    log.warn("{} event ignored", __func__);
}

void query_processor::migration_subscriber::on_drop_view(const sstring& ks_name, const sstring& view_name) {
    remove_invalid_prepared_statements(ks_name, view_name);
}
```

nyh reviewed (changes requested):
Looks mostly good, but I left a few small comments/requests.

avikivity commented on test/cql/where_clause_test.cql:
```diff
@@ -0,0 +1,18 @@
+-- setup
+CREATE KEYSPACE ks1 WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1}
+        AND DURABLE_WRITES = true;
+ALTER KEYSPACE ks1 WITH replication = { 'class' : 'NetworkTopologyStrategy', 'datacenter1' : 1 };
```
Probably cut'n'paste

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -2648,5 +2648,25 @@ bool is_partition_token_for_schema(const expression& maybe_token, const schema&
+        return std::visit(find_agg, fc.func);
+    });
+    if (found_agg) {
+        throw exceptions::invalid_request_exception(fmt::format("Aggregation function are not supported in the {}", context_for_errors));
```
It's because I don't want the 20-patch series to fail because some dtest somewhere checks for this message.  Grammar fixes can be done in independent patches elsewhere.

nyh replied:
This is true, but a day after you send this PR, nobody will ever remember to fix the message string :-(

At least it's sad to repeat the same mistake you said dtest is doing in the test you added yourself.

Ideally, I would like to see dtest (and your test) check partial error strings, not entire English sentences.

nyh commented on cql3/expr/restrictions.cc:
```diff
@@ -184,6 +184,7 @@ binary_operator validate_and_prepare_new_restriction(const binary_operator& rest
 
     // Prepare the restriction
     binary_operator prepared_binop = prepare_binary_operator(restriction, db, *schema);
+    expr::verify_no_aggregate_functions(prepared_binop, "where clause");
```
Ok, I didn't know that there was an existing error message with that string.
Sad that dtest, instead of detecting bugs (in this case, English bugs), forces us to keep them unfixed :-(

avikivity commented:
v3:
 - keep field name for selectable::with_field_selection::to_string()
 - switched test for rejection of aggregates in the WHERE clause to pytest
 - dropped unneeded/merged patches
 - rebased

avikivity commented:
@nyh I tried the "dismiss review" but it just killed all your review comments; I thought it would explain that I addressed the requests raised.

avikivity commented:
Of it didn't kill the comments, just the "request changes".

avikivity commented:
I'll fix that bug as soon as it promotes.

Note it wasn't even dtest preventing the bug fix - it was fear of a dtest preventing the bug fix. But in a standalone fix there's no risk of derailing a large series over a triviality.
